tensorboard/backend/security_validator.py

```diff
@@ -101,56 +115,81 @@ def _validate_csps(self, headers):
-        if rule in _CSP_IGNORE.get(name, []):
-          # There are cases where certain values are legitimate.
-          continue
+      if value in _CSP_IGNORE.get(name, []):
+        # There are cases where certain values are legitimate.
+        continue
 
-        if (
-            rule == "'self'" or rule == "'none'" or rule.startswith("https:")
-            or rule.startswith("'sha256-")
-        ):
-          continue
+      # TensorBoard follows principle of least privilege. However, to make it
+      # easier to conform to the security policy for plugin authors, TensorBoard
+      # trusts request and resources originating its server. Also, it can
+      # selectively trust domains as long as they use https protocol. Lastly, it
+      # can allow 'none' directive.
+      # TODO(stephanwlee): deprecate the sha-based whitelisting.
+      if (
+          value == "'self'" or value == "'none'" or value.startswith("https:")
+          or value.startswith("'sha256-")
+      ):
+        continue
 
-        msg = "Illegal Content-Security-Policy for {name}: {rule}".format(
-            name=name, rule=rule
-        )
-        violations.append(msg)
+      msg = "Illegal Content-Security-Policy for {name}: {value}".format(
+          name=name, value=value
+      )
+      violations.append(msg)
 
     if not has_default_src:
       violations.append("Requires default-src for Content-Security-Policy")
 
     if violations:
-      self._maybe_raise_value_error("\n".join(violations))
+      _maybe_raise_value_error("\n".join(violations))
 
-  def _parse_csp_text(self, csp_text):
+  def _parse_serialized_csp(self, csp_text):
+    # See https://www.w3.org/TR/CSP/#parse-serialized-policy.
+    # Below Steps are based on the algorithm stated in above spec.
+    # Deviations:
+    # - it does not warn and ignore duplicative directive (Step 2.5)
+
+    # Step 2
     csp_srcs = csp_text.split(";")
+
     policies = []
     for csp_src in csp_srcs:
+      # Step 2.1
+      csp_src = csp_src.strip()
+
+      if not csp_src:
+        # Step 2.2
+        continue
+
+      # Step 2.3
       match = _CSP_PATTERN.match(csp_src)
-      if not match or len(match.groups()) < 2:
+      if not match or len(match.groups()) < 2 or not match.group(1).strip():
         continue
-      name, rules = _CSP_PATTERN.match(csp_src).groups()
-      policies.append(
-          Policy(
-              name=name,
-              rules=[rule.strip() for rule in rules.split(" ") if rule]
-          )
-      )
 
-    return policies
+      name, values = _CSP_PATTERN.match(csp_src).groups()
+      # Step 2.4
+      name = name.lower()
 
-  def _maybe_raise_value_error(self, error_msg):
-    logger.warn("In 3.0, this warning will become an error:\n%s" % error_msg)
-    # TODO(3.x): raise a value error.
+      # Step 2.6
+      for value in values.split(" "):
+        if not value:
+          continue
+
+        # Step 2.7
+        directive = Directive(name=name, value=value)
```

Comment:
This isn't exactly the step as described in the spec - there should be a single directive emitted with the name `name` for which the value of the directive is a set of tokens.  So basically just `policies.append(Directive(name=name, value=values.split()))`.

Also a nit but I think it should be `split()` not `split(" ")` to split on chunks of whitespace rather than only on spaces (and then we don't have to filter out empty strings in the case of consecutive spaces either).